starky/src/prover.rs

```diff
@@ -142,6 +177,81 @@ where
     })
 }
 
+/// Compute all Z polynomials (for permutation arguments).
+fn compute_z_polys<F, C, S, const D: usize>(
+    stark: &S,
+    config: &StarkConfig,
+    challenger: &mut Challenger<F, C::Hasher>,
+    trace_poly_values: &[PolynomialValues<F>],
+) -> Vec<PolynomialValues<F>>
+where
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    S: Stark<F, D>,
+{
+    let permutation_pairs = stark.permutation_pairs();
+    let permutation_challenge_sets = get_n_permutation_challenge_sets(
+        challenger,
+        config.num_challenges,
+        stark.permutation_batch_size(),
+    );
+
+    let permutation_instances = permutation_pairs
+        .iter()
+        .cartesian_product(0..config.num_challenges)
+        .chunks(stark.permutation_batch_size())
+        .into_iter()
+        .flat_map(|batch| {
+            batch.enumerate().map(|(i, (pair, chal))| {
+                let challenge = permutation_challenge_sets[i].challenges[chal];
+                PermutationInstance { pair, challenge }
+            })
+        })
+        .collect_vec();
+
+    permutation_instances
+        .into_par_iter()
+        .map(|instance| compute_z_poly(instance, trace_poly_values))
+        .collect()
+}
+
+/// Compute a single Z polynomial.
+fn compute_z_poly<F: Field>(
+    instance: PermutationInstance<F>,
+    trace_poly_values: &[PolynomialValues<F>],
+) -> PolynomialValues<F> {
+    let PermutationInstance { pair, challenge } = instance;
+    let PermutationPair {
+        lhs_columns,
+        rhs_columns,
+    } = pair;
+    let PermutationChallenge { beta, gamma } = challenge;
+
+    let degree = trace_poly_values[0].len();
+    let mut reduced_lhs = PolynomialValues::constant(gamma, degree);
+    let mut reduced_rhs = PolynomialValues::constant(gamma, degree);
+
+    let both_cols = lhs_columns.iter().zip_eq(rhs_columns);
+    for ((lhs, rhs), weight) in both_cols.zip(beta.powers()) {
+        reduced_lhs.add_assign_scaled(&trace_poly_values[*lhs], weight);
+        reduced_rhs.add_assign_scaled(&trace_poly_values[*rhs], weight);
+    }
+
+    // Compute the quotients.
+    let reduced_rhs_inverses = F::batch_multiplicative_inverse(&reduced_rhs.values);
+    let mut quotients = reduced_lhs.values;
+    batch_multiply_inplace(&mut quotients, &reduced_rhs_inverses);
+
+    // Compute Z, which contains partial products of the quotients.
+    let mut partial_products = Vec::with_capacity(degree);
+    let mut acc = F::ONE;
+    for q in quotients {
+        partial_products.push(acc);
+        acc *= q;
+    }
+    PolynomialValues::new(partial_products)
+}
```

Comment:
Maybe I'm missing something, but it seems like this could be done over batches of `quotient_degree_factor` instances.

If I understand correctly, this does (in the case of two lhs columns `l_0, l_1` and two rhs columns `r_0, r_1`):
```
Z(gx) = Z(x) * (gamma + l_0 + beta * l_1) / (gamma + r_0 + beta * r_1)
=>
Z(gx) * (gamma + r_0 + beta * r_1) - Z(x) * (gamma + l_0 + beta * l_1) = 0.
```
This last constraint is of degree 2, whereas the vanishing polynomial can be of degree up to 3 (with `rate_bits=1`). 

So it seems like having `Z` take batches of `quotient_degree_factor` instances would yield a constraint of degree `constraint_degree`, and would result in fewer `Z` polynomials.

Reply:
Ah yes, that was my intention but I forgot to actually implement the batching logic here. I'll just merge the non-batch version for now, as I'm not sure when I'll have time to finish this and want to minimize conflicts, but I added a TODO and I'll come back to this when I have a chance.

Reply:
Sounds good 👍